presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -360,20 +568,159 @@ protected Page computeNext()
                             BlockBuilder output = pageBuilder.getBlockBuilder(types.size() + i);
                             aggregator.evaluate(groupId, output);
                         }
-
-                        groupId++;
                     }
 
                     return pageBuilder.build();
                 }
             };
         }
+
+        private void rebuildAggregators()
+        {
+            groupByHash = createGroupByHash();
+            aggregators = createAggregators();
+        }
+
+        private void rebuildAggregatorsForMerge()
+        {
+            groupByHash = createGroupByHash();
+            aggregators = createAggregatorsForMerge(groupByHash.getTypes().size());
+        }
+
+        private GroupByHash createGroupByHash()
+        {
+            return GroupByHash.createGroupByHash(operatorContext.getSession(), groupByTypes, Ints.toArray(groupByChannels), hashChannel, expectedGroups);
+        }
+
+        private List<Aggregator> createAggregators()
+        {
+            // wrapper each function with an aggregator
+            ImmutableList.Builder<Aggregator> builder = ImmutableList.builder();
+
+            requireNonNull(accumulatorFactories, "accumulatorFactories is null");
+            for (int i = 0; i < accumulatorFactories.size(); i++) {
+                AccumulatorFactory accumulatorFactory = accumulatorFactories.get(i);
+                builder.add(new Aggregator(accumulatorFactory, step));
+            }
+            return builder.build();
+        }
+
+        private List<Aggregator> createAggregatorsForMerge(int numberOfColumnsInKey)
```

Comment:
because we need it to determine channel number (`intermediateChannel`) for each aggregator. It could be renamed to something like `intermediateChannelStartsAt` but I don't know if that it's better.